src/snapred/backend/data/Indexor.py

```diff
@@ -0,0 +1,367 @@
+        """
+        Path to a specific version of a calculation record
+        """
+        return self.versionPath(version) / f"{self.indexorType}Record.json"
+
+    def parametersPath(self, version: Optional[int] = None):
+        """
+        Path to a specific version of calculation parameters
+        """
+        return self.versionPath(version) / f"{self.indexorType}Parameters.json"
+
+    @validate_call
+    def versionPath(self, version: Optional[int] = None) -> Path:
+        if version is None:
+            version = VERSION_START
+        else:
+            version = self.thisOrCurrentVersion(version)
+        return self.rootDirectory / wnvf.fileVersion(version)
+
+    def currentPath(self) -> Path:
+        """
+        The latest version of those in the index,
+        or the version you just added to the index.
+        """
+        return self.versionPath(self.currentVersion())
+
+    def latestApplicablePath(self, runNumber: str) -> Path:
+        return self.versionPath(self.latestApplicableVersion(runNumber))
+
+    ## INDEX MANIPULATION METHODS ##
+
+    def getIndex(self) -> List[IndexEntry]:
+        if self.index == {}:
+            self.index = self.readIndex()
+
+        # remove the default version, if it exists
+        res = self.index.copy()
+        res.pop(VERSION_DEFAULT, None)
+        return list(res.values())
+
+    def readIndex(self) -> Dict[int, IndexEntry]:
+        # create the index from the index file
+        indexPath: Path = self.indexPath()
+        indexList: List[IndexEntry] = []
+        if indexPath.exists():
+            indexList = parse_file_as(List[IndexEntry], indexPath)
+        return {entry.version: entry for entry in indexList}
+
+    def writeIndex(self):
+        path = self.indexPath()
+        path.parent.mkdir(parents=True, exist_ok=True)
+        write_model_list_pretty(self.index.values(), path)
+
+    def addIndexEntry(self, entry: IndexEntry):
+        """
+        Will save at the version on the index entry.
+        If the version is invalid, will throw an error and refuse to save.
+        """
+        if not self.isValidVersion(entry.version):
+            raise RuntimeError(f"Invalid version {entry.version} on index entry.  Save failed.")
+
+        self.index[entry.version] = entry
+        self.writeIndex()
+
+    def indexEntryFromRecord(self, record: Record) -> IndexEntry:
+        return Record.indexEntryFromRecord(record)
+
+    ## RECORD READ / WRITE METHODS ##
+
+    def readRecord(self, version: Optional[int] = None) -> Record:
+        """
+        If no version given, defaults to current version
+        """
+        version = self.thisOrCurrentVersion(version)
+        filePath = self.recordPath(version)
+        record = None
+        if filePath.exists():
+            match self.indexorType:
+                case IndexorType.CALIBRATION:
+                    record = parse_file_as(CalibrationRecord, filePath)
```

Comment:
could the record class not also be an init param?  It would save you these case statements.

Reply:
Are you okay with
``` python
record = parse_file_as(globals()[f"{self.indexorType}Record"], filePath)
```
?

Reply:
It works without the `match`, but it's ugly and not as perspicuous